File: src/main/java/net/rakugakibox/util/YamlResourceBundle.java
package net.rakugakibox.util;

import java.io.IOException;
import java.io.InputStream;
import java.io.Reader;
import java.util.AbstractMap.SimpleImmutableEntry;
import static java.util.Arrays.asList;
import static java.util.Collections.enumeration;
import static java.util.Collections.unmodifiableList;
import java.util.Enumeration;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Map.Entry;
import java.util.ResourceBundle;
import java.util.Set;
import static java.util.stream.Collectors.toMap;
import java.util.stream.Stream;

import org.yaml.snakeyaml.Yaml;

import lombok.Cleanup;
import lombok.NonNull;

/**
 * {@link ResourceBundle} for YAML format.
 */
public class YamlResourceBundle extends ResourceBundle {

    /**
     * Entries.
     */
    private final Map<String, Object> entries;

    /**
     * Constructor.
     *
     * @param string YAML data.
     */
    public YamlResourceBundle(@NonNull String string) {
        entries = flattenYamlTree(new Yaml().loadAs(string, Map.class));
    }

    /**
     * Constructor.
     *
     * @param stream YAML data as input stream.
     */
    public YamlResourceBundle(@NonNull InputStream stream) {
        entries = flattenYamlTree(new Yaml().loadAs(stream, Map.class));
    }

    /**
     * Constructor.
     *
     * @param reader YAML data as input character stream.
     */
    public YamlResourceBundle(@NonNull Reader reader) {
        entries = flattenYamlTree(new Yaml().loadAs(reader, Map.class));
    }

    /**
     * Flatten yaml tree structure.
     *
     * @param map {@link Map} of yaml tree.
     * @return {@link Map} of entries.
     */
    private static Map<String, Object> flattenYamlTree(Map<?, ?> map) {
        return map.entrySet().stream()
                .flatMap(YamlResourceBundle::flattenYamlTree)
                .collect(toMap(
                        e -> e.getKey(),
                        e -> e.getValue(),
                        (oldValue, newValue) -> newValue
                ));
    }

    /**
     * Flatten yaml tree structure.
     *
     * @param entry {@link Entry} of yaml tree.
     * @return {@link Stream} of entries.
     */
    private static Stream<Entry<String, Object>> flattenYamlTree(Entry<?, ?> entry) {
        String key = entry.getKey().toString();
        Object value = entry.getValue();
        if (value instanceof Map) {
            Map<?, ?> valueAsMap = (Map<?, ?>) value;
            return valueAsMap.entrySet().stream()
                    .flatMap(YamlResourceBundle::flattenYamlTree)
                    .map(e -> new SimpleImmutableEntry<>(key + "." + e.getKey(), e.getValue()));
        } else if (value instanceof List) {
            List<?> valueAsList = (List<?>) value;
            value = valueAsList.stream().toArray(String[]::new);
        }
        return Stream.of(new SimpleImmutableEntry<>(key, value));
    }

    /** {@inheritDoc} */
    @Override
    protected Set<String> handleKeySet() {
        return entries.keySet();
    }

    /** {@inheritDoc} */
    @Override
    public Enumeration<String> getKeys() {
        return enumeration(keySet());
    }

    /** {@inheritDoc} */
    @Override
    protected Object handleGetObject(@NonNull String key) {
        return entries.get(key);
    }

    /**
     * {@link ResourceBundle.Control} for YAML format.
     */
    public static class Control extends ResourceBundle.Control {

        /**
         * Singleton instance.
         */
        public static final Control INSTANCE = new Control();

        /**
         * Constructor.
         */
        protected Control() {
        }

        /** {@inheritDoc} */
        @Override
        public List<String> getFormats(@NonNull String baseName) {
            return unmodifiableList(asList("yaml", "yml"));
        }

        /** {@inheritDoc} */
        @Override
        public ResourceBundle newBundle(@NonNull String baseName,
                @NonNull Locale locale, @NonNull String format, @NonNull ClassLoader loader, boolean reload)
                throws IllegalAccessException, InstantiationException, IOException {
            if (!getFormats(baseName).contains(format)) {
                return null;
            }
            String bundleName = toBundleName(baseName, locale);
            String resourceName = toResourceName(bundleName, format);
            @Cleanup InputStream stream = loader.getResourceAsStream(resourceName);
            return new YamlResourceBundle(stream);
        }

    }

}

File: src/main/kotlin/dev/akkinoc/util/YamlResourceBundle.kt
package dev.akkinoc.util

import org.yaml.snakeyaml.Yaml
import java.io.InputStream
import java.io.Reader
import java.util.Collections.enumeration
import java.util.Enumeration
import java.util.Locale
import java.util.ResourceBundle

/**
 * [ResourceBundle] for YAML format.
 *
 * @property entries The resource bundle entries.
 */
class YamlResourceBundle private constructor(private val entries: Map<String, Any>) : ResourceBundle() {

    constructor() : this(emptyMap())

    /**
     * @param docs A string containing the YAML documents.
     */
    constructor(docs: String) : this(Parser.parseDocs(docs))

    /**
     * @param docs An input stream to read the YAML documents.
     */
    constructor(docs: InputStream) : this(Parser.parseDocs(docs))

    /**
     * @param docs A reader to read the YAML documents.
     */
    constructor(docs: Reader) : this(Parser.parseDocs(docs))

    override fun getKeys(): Enumeration<String> = enumeration(keySet())

    override fun handleKeySet(): Set<String> = entries.keys

    override fun handleGetObject(key: String): Any? = entries[key]

    /**
     * [ResourceBundle.Control] for YAML format.
     */
    object Control : ResourceBundle.Control() {

        override fun getFormats(baseName: String): List<String> = listOf("yaml", "yml")

        override fun newBundle(
            baseName: String,
            locale: Locale,
            format: String,
            loader: ClassLoader,
            reload: Boolean,
        ): ResourceBundle? {
            require(format in getFormats(baseName)) { "Unknown format: $format" }
            val bundleName = toBundleName(baseName, locale)
            val resourceName = toResourceName(bundleName, format)
            val resource = loader.getResource(resourceName) ?: return null
            val connection = resource.openConnection()
            if (reload) connection.useCaches = false
            return connection.getInputStream().use { YamlResourceBundle(it) }
        }

    }

    /**
     * The YAML documents parser.
     */
    private object Parser {

        /**
         * Parses the YAML documents.
         *
         * @param docs A string containing the YAML documents.
         * @return The resource bundle entries.
         */
        fun parseDocs(docs: String): Map<String, Any> = parseDocs { Yaml().loadAll(docs) }

        /**
         * Parses the YAML documents.
         *
         * @param docs An input stream to read the YAML documents.
         * @return The resource bundle entries.
         */
        fun parseDocs(docs: InputStream): Map<String, Any> = parseDocs { Yaml().loadAll(docs) }

        /**
         * Parses the YAML documents.
         *
         * @param docs A reader to read the YAML documents.
         * @return The resource bundle entries.
         */
        fun parseDocs(docs: Reader): Map<String, Any> = parseDocs { Yaml().loadAll(docs) }

        /**
         * Parses the YAML documents.
         *
         * @param read A function to read the YAML documents.
         * @return The resource bundle entries.
         */
        private fun parseDocs(read: () -> Iterable<Any>): Map<String, Any> {
            val docs = read()
            val entries = docs.asSequence().flatMap { parseDoc(it) }.toMap()
            return entries.asSequence().mapNotNull { (key, value) -> value?.let { key to it } }.toMap()
        }

        /**
         * Parses the YAML document.
         *
         * @param doc The YAML document.
         * @return The resource bundle entries.
         */
        private fun parseDoc(doc: Any): Sequence<Pair<String, Any?>> {
            require(doc is Map<*, *>) { "The root of the YAML document must be a map." }
            return doc.asSequence().flatMap { (itemKey, itemValue) ->
                parseNode(key = "$itemKey", value = itemValue, ancestors = emptyList())
            }
        }

        /**
         * Parses the YAML document node.
         *
         * @param key The key of the node in the YAML document.
         * @param value The value of the node in the YAML document.
         * @param ancestors A list of node values from root to parent.
         * @return The resource bundle entries.
         */
        private fun parseNode(key: String, value: Any?, ancestors: List<Any>): Sequence<Pair<String, Any?>> {
            return when (value) {
                is Map<*, *> -> parseMapNode(key = key, value = value, ancestors = ancestors)
                is List<*> -> parseListNode(key = key, value = value, ancestors = ancestors)
                else -> sequenceOf(key to value)
            }
        }

        /**
         * Parses the YAML document map node.
         *
         * @param key The key of the node in the YAML document.
         * @param value The value of the node in the YAML document.
         * @param ancestors A list of node values from root to parent.
         * @return The resource bundle entries.
         */
        private fun parseMapNode(key: String, value: Map<*, *>, ancestors: List<Any>): Sequence<Pair<String, Any?>> {
            if (ancestors.any { it === value }) return emptySequence()
            val path = ancestors + value as Any
            return value.asSequence().flatMap { (itemKey, itemValue) ->
                parseNode(key = "$key.$itemKey", value = itemValue, ancestors = path)
            }
        }

        /**
         * Parses the YAML document list node.
         *
         * @param key The key of the node in the YAML document.
         * @param value The value of the node in the YAML document.
         * @param ancestors A list of node values from root to parent.
         * @return The resource bundle entries.
         */
        private fun parseListNode(key: String, value: List<*>, ancestors: List<Any>): Sequence<Pair<String, Any?>> {
            if (ancestors.any { it === value }) return emptySequence()
            val path = ancestors + value as Any
            val strings = value.run {
                if (isEmpty()) return@run emptySequence()
                val items = map { it as? String ?: return@run emptySequence() }
                sequenceOf(key to items.toTypedArray())
            }
            val items = value.asSequence().flatMapIndexed { itemIndex, itemValue ->
                parseNode(key = "$key[$itemIndex]", value = itemValue, ancestors = path)
            }
            return strings + items
        }

    }

}
